### PR TITLE
Feat/127 groups repo

### LINKED ISSUE #127: Groups Repository
As a user, I want the app to manage group information so that I can join, leave, and interact with different groups.

### PR DESCRIPTION
**Pull request by AACioffi**
## What?
Added the `GroupsRepositoryFirestore` class - a complete Firestore implementation of the `GroupsRepository` interface. This includes full CRUD operations, member/admin management, and security enforcement for group operations. Also includes a comprehensive test suite with 25 tests covering all functionality and edge cases. Closes #127.

## Why?
This completes the repository layer for the groups feature, providing the actual persistence layer that interacts with Firebase Firestore. The implementation enforces all business rules and security constraints (admin-only operations, self-removal permissions, last-admin protection) at the repository level, ensuring data integrity regardless of how the repository is called.

## How?
**Implementation (`GroupsRepositoryFirestore`):**
- Uses Firebase Firestore with a `groups` collection, along with dedicated rules
- Implements all 11 interface methods with proper async/await patterns
- Enforces security rules inline: checks current user's admin status before operations, allows self-removal for members, prevents removing the last admin
- Automatic creator initialization: ensures group creators are added as both members and admins
- Helper methods `snapshotToGroup()` and `groupToMap()` for Firestore serialization/deserialization

**Testing (`GroupsRepositoryFirestoreTest` + `FirestoreGroupsGatherlyTest`):**
- 25 comprehensive tests covering CRUD operations, security enforcement, member/admin management, and edge cases
- Uses Firebase Emulator Suite with fake JWT tokens for multi-user scenarios
- Tests 4 different users to verify permissions and group membership isolation
- Base class `FirestoreGroupsGatherlyTest` handles emulator setup/teardown and user authentication

## Testing?
Full test coverage included with emulator-based integration tests, covering:
- Basic CRUD operations
- Security: admin-only edit/delete/add-member/remove-member/add-admin/remove-admin
- Member self-removal (leave group)
- Last admin prevention on both removal and demotion
- Multi-user scenarios and group isolation
- Edge cases

## Screenshots
image
All 25 tests of the suite pass locally.

## Anything Else?
The `signInWithToken()` helper method in `FirestoreGroupsGatherlyTest` is currently duplicated across multiple test auxiliary classes. This could be refactored into a shared test utility file to reduce code duplication.

The test suite is intentionally comprehensive (~400 lines) to ensure as many security rules and edge cases as possible are properly validated before merging. Keep in mind this constitutes the bulk of the PR.

Note that this PR includes the `Group` data class, as well as the `GroupsRepository` interface, as seen in PRs #167 and #168. If they have already been approved, you may forego reviewing them in detail in this PR.

### CODE REVIEW

File: app/src/main/java/com/android/gatherly/model/group/Group.kt
Comment (clairechaffard) [inline app/src/main/java/com/android/gatherly/model/group/Group.kt]: Positive: Already been approved in a previous PR, 
Important: just be careful when you will merge your different PR.

File: app/src/androidTest/java/com/android/gatherly/model/group/GroupsRepositoryFirestoreTest.kt [hunk 1]
```diff
@@ -0,0 +1,392 @@
...
+
+    // Sign in as user4 (not an admin)
+    signInWithToken(user4Token)
+    try {
+      repository.deleteGroup(testedGroup.gid)
+      fail("Expected SecurityException to be thrown")
+    } catch (e: SecurityException) {
+      // Expected
+    }
+  }
```
Comment (thisshrek) [inline app/src/androidTest/java/com/android/gatherly/model/group/GroupsRepositoryFirestoreTest.kt]: Optional : It seems like this test doesn't fail if the exception thrown is something other than a security exception, we can optionally add a case for all the other exceptions, in which the test fails.

Same for all other tests in this file that check for a security exception.

File: app/src/androidTest/java/com/android/gatherly/model/group/GroupsRepositoryFirestoreTest.kt [hunk 2]
```diff
@@ -0,0 +1,392 @@
...
+  @Test
+  fun removeMember_prevents_removing_last_admin() = runTest {
+    val testedGroup = group2.copy(memberIds = listOf(user1Id, user2Id), adminIds = listOf(user1Id))
+    repository.addGroup(testedGroup)
+    try {
+      repository.removeMember(testedGroup.gid, user1Id)
+      fail("Expected IllegalStateException to be thrown")
+    } catch (e: IllegalStateException) {
+      // Expected
+    }
```
Comment (thisshrek) [inline app/src/androidTest/java/com/android/gatherly/model/group/GroupsRepositoryFirestoreTest.kt]: Optional : Same here, this test doesn't fail if the exception thrown is something other than an illegal state exception, we can optionally add a case for all the other exceptions, in which the test fails.

Same for all other tests in this file that check for an illegal state exception.

File: app/src/androidTest/java/com/android/gatherly/model/group/GroupsRepositoryFirestoreTest.kt [hunk 3]
```diff
@@ -0,0 +1,392 @@
...
+      repository.addAdmin(testedGroup.gid, user3Id)
+      fail("Expected SecurityException to be thrown")
+    } catch (e: SecurityException) {
+      // Expected
+    }
+  }
+
+  @Test
+  fun removeAdmin_removes_user_from_adminIds() = runTest {
+    val testedGroup =
```
Comment (thisshrek) [inline app/src/androidTest/java/com/android/gatherly/model/group/GroupsRepositoryFirestoreTest.kt]: Optional : Since you documented most of the tests in this file, There are a few tests which lack documentation. Feel free to document them in case this wasn't intentional, but it is not mandatory at all since the test names should be self explanatory and sufficient.

**Review by thisshrek — APPROVED** [review]
Very impressive PR, by far one of the best I've reviewed 💯 . Excellent structure and completeness!

You’ve implemented a secure, consistent, and well tested repository. It covers full CRUD, admin/member rules, and safety checks a with strong and VERY thorough testing.

I've left some optional suggestions and a question, but these are non blockers and therefore I'll approve the PR. Nevertheless, please make sure to have a look at my comments and address them before you merge. Thanks!

Good job👍🏼 !

File: app/src/androidTest/java/com/android/gatherly/utils/FirestoreGroupsGatherlyTest.kt [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+  protected lateinit var user1Id: String
+  protected lateinit var user2Id: String
+  protected lateinit var user3Id: String
+  protected lateinit var user4Id: String
+
+  /** Sample group used as a template in tests. Created by Alice. */
+  protected val group1 =
+      Group(
+          gid = "1",
+          creatorId = "", // Will be set dynamically in tests
```
[the remark this answers is not in the document]
Comment (AACioffi) [reply app/src/androidTest/java/com/android/gatherly/utils/FirestoreGroupsGatherlyTest.kt]: The creatorIds are set during tests, as indicated by the comment. This is done through addGroup().

File: app/src/androidTest/java/com/android/gatherly/model/group/GroupsRepositoryFirestoreTest.kt (hunk 1 above)
Comment (AACioffi) [reply app/src/androidTest/java/com/android/gatherly/model/group/GroupsRepositoryFirestoreTest.kt]: The tests (or most of them) follow the given pattern:
- SecurityException thrown -> caught -> test passes
- Different exception thrown -> propagates -> test fails
- No exception thrown -> fail() -> test fails

It seems to me that all cases are covered. But I get that one would prefer, say, assertThrows() or something similar. It would specifically fail the test if SecurityException isn't thrown. I will keep that in mind for future tests, and more consistently use this kind of assertion.
